File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/engine/soul/HeadlineSyncClient.kt
package org.wyrdsekai.app.engine.soul

import kotlinx.serialization.Serializable

/**
 * Headline — a lightweight status broadcast between buds of the same lineage.
 *
 * Headlines are the continuous (~200B) sync mechanism from §95 (Soul Budding).
 * Each bud periodically posts a headline summarizing its state; siblings consume
 * these to maintain awareness of the family without full sync.
 *
 * Human sees one companion; buds coordinate via headlines + warm handoff.
 */
@Serializable
data class Headline(
    val budDid: String,
    val summary: String,
    val vitalitySnapshot: Map<String, Float>,
    val itemCount: Int,
    val timestamp: Long,
)

/**
 * Client-side interface for bud headline synchronization.
 *
 * Posting and receiving will be wired to the Between (WebSocket/NATS)
 * once that transport layer is integrated. Until then, the in-memory
 * implementation serves for local testing and single-device use.
 */
interface HeadlineSyncClient {
    /** Post this bud's current headline to the family. */
    suspend fun postHeadline(headline: Headline)

    /** Return the latest cached headlines from all family buds. */
    fun latestHeadlines(): List<Headline>

    /** Register a callback invoked when a headline arrives from a sibling. */
    fun onHeadlineReceived(callback: (Headline) -> Unit)
}

/**
 * In-memory headline sync — caches headlines locally, no network.
 * Useful for single-bud testing and phone-only mode.
 */
class InMemoryHeadlineSyncClient : HeadlineSyncClient {
    private val headlines = mutableMapOf<String, Headline>()
    private val listeners = mutableListOf<(Headline) -> Unit>()

    override suspend fun postHeadline(headline: Headline) {
        headlines[headline.budDid] = headline
        listeners.forEach { it(headline) }
    }

    override fun latestHeadlines(): List<Headline> = headlines.values.toList()

    override fun onHeadlineReceived(callback: (Headline) -> Unit) {
        listeners.add(callback)
    }
}

File: clients/rn/src/engine/soul/HeadlineSyncClient.ts
/**
 * Headline sync client for soul buds (§95).
 *
 * Buds continuously exchange lightweight "headline" summaries (~200B)
 * so family members stay loosely coordinated without full sync.
 * On phones, this runs over the Between (NATS) or degrades to polling.
 *
 * Headlines are ephemeral — only the latest per bud is retained.
 */

export interface Headline {
  budDid: string;
  summary: string;
  vitalitySnapshot: Record<string, number>;
  itemCount: number;
  timestamp: number;
}

export type HeadlineCallback = (headline: Headline) => void;

export class HeadlineSyncClient {
  private headlines = new Map<string, Headline>();
  private listeners: HeadlineCallback[] = [];

  /**
   * Post a headline from this bud. Stores locally and notifies listeners.
   * In production, this would also publish to the Between (NATS topic).
   */
  postHeadline(headline: Headline): void {
    this.headlines.set(headline.budDid, headline);
    for (const listener of this.listeners) {
      listener(headline);
    }
  }

  /**
   * Receive a headline from a remote bud (e.g., via Between/NATS subscription).
   * Stores and notifies listeners.
   */
  receiveHeadline(headline: Headline): void {
    this.headlines.set(headline.budDid, headline);
    for (const listener of this.listeners) {
      listener(headline);
    }
  }

  /**
   * Get the latest headlines from all known buds, sorted by timestamp (newest first).
   */
  latestHeadlines(): Headline[] {
    return [...this.headlines.values()].sort((a, b) => b.timestamp - a.timestamp);
  }

  /**
   * Get the latest headline for a specific bud DID.
   */
  headlineFor(budDid: string): Headline | null {
    return this.headlines.get(budDid) ?? null;
  }

  /**
   * Subscribe to incoming headlines. Returns unsubscribe function.
   */
  onHeadlineReceived(callback: HeadlineCallback): () => void {
    this.listeners.push(callback);
    return () => {
      this.listeners = this.listeners.filter(l => l !== callback);
    };
  }

  /**
   * Clear all stored headlines.
   */
  clear(): void {
    this.headlines.clear();
  }
}
